src/properties.c

```diff
@@ -0,0 +1,683 @@
+/*
+#define DEBUG
+ */
+
+#include "vipsdisp.h"
+
+/* The Properties widget displays the properties of the VipsImage. This includes
+ * geometric values such as width and height as well as any EXIF metadata.
+ */
+
+/* This structure defines the properties of a Properties object.
+ * (GObject boilerplate)
+ */
+struct _Properties
+{
+	GtkWidget parent_instance;
+
+	ImageWindow *image_window;
+
+	GtkWidget *scrolled_window;
+	GtkWidget *search_entry;
+	GtkWidget *search_warning;
+
+	GtkWidget *main_box;
+	GtkWidget *left_column;
+	GtkWidget *right_column;
+	GList *field_list;
+	int field_list_length;
+	gboolean ignore_case;
+};
+
+/* This macro defines the Properties type.
+ * (GObject boilerplate)
+ */
+G_DEFINE_TYPE( Properties, properties, GTK_TYPE_WIDGET );
+
+/* This enum defines the signals used to get and set custom properties.
+ * See:
+ * 	- properties_get_property
+ * 	- properties_set_property
+ *
+ * (GObject boilerplate).
+ */
+enum {
+	PROP_IMAGE_WINDOW = 1,
+	PROP_REVEALED,
+
+	SIG_LAST
+};
+
+/* Clean up any existing children of @left_column and @right_column and any
+ * existing elements of the GList @field_list.
+ *
+ * @m	Properties *	this
+ */
+static void
+properties_clear_main_box( Properties *p )
+{
+	GtkWidget *t0, *t1;
+
+	t0 = gtk_widget_get_last_child( p->left_column );
+	while ( t0 ) {
+		t1 = gtk_widget_get_prev_sibling( t0 );
+		properties_util_free_label_box( t0 );
+		t0 = t1;
+	}
+
+	t0 = gtk_widget_get_last_child( p->right_column );
+	while ( t0 ) {
+		t1 = gtk_widget_get_prev_sibling( t0 );
+		properties_util_free_input_box( t0 );
+		t0 = t1;
+	}
+
+	if( p->field_list ) {
+		g_list_free( p->field_list );
```

Comment:
Should this be `g_list_free_full()`?

Reply:
`g_list_free_full` isn't what we want here, since `g_strfreev` cleans up the dynamically allocated `fields`, at the end of `properties_create_main_box( Properties *p )` (line 145). All that is left its the GList itself, which is cleaned up with `g_list_free`. Not super obvious.

The only copies of that data that are made are by the label widgets themselves, internally. `gtk_label_get` returns a const pointer to its copy, and the GtkLabel cleans it up when it goes away. When I call `gtk_label_set_label` and pass a dynamically allocated pointer (e.g., a pointer in `fields`), I am still responsible for freeing it. That just happens conveniently all with the one `g_strfreev` call, since I was careful not to make extra copies.

Reply:
At the moment you're freeing the strings and the array with `g_strfreev()`, but the pointers to the strings are still in `p->field_list`. You'll get a crash when the memory is reused.

I think I would steal the strings for the list and just free the spine of the array with `g_free()`. But then you'll need `g_list_free_full()`. The alternative would be to `g_strdup()` the strings into the list.